### ISSUE TITLE
Should preactions run before dbtable creation?

### ISSUE THREAD

**Issue by bryan0** [databricks/spark-redshift#268]
RedshiftWriter in overwrite mode create a new schema based on the dataset.  This can be a problem if you want to guarantee the same redshift table schema and fail if there is a an inconsistency.  This is also a problem because of the varchar(256) issue (https://github.com/databricks/spark-redshift/issues/118) and metadata modification isn't even supported outside of scala.

to achieve this in redshift you might do something like:
BEGIN;
CREATE temp_my_table (LIKE my_table);
-- load into temp_my_table
DROP TABLE my_table;
ALTER TABLE temp_my_table RENAME TO my_table;
END;

I tried to have this same behavior with spark-redshift using append mode with something like:

ds.write
....
.option("preactions", "CREATE TABLE temp_my_table (LIKE my_table);")
.option("dbtable", "my_temp_table")
.option("postctions", "DROP TABLE my_table; ALTER TABLE my_temp_table RENAME TO my_table;")
.mode("append")
.save()

but this fails because RedshiftWriter creates the dbtable before running the preaction.

Should the dbtable creation really happen before the preaction?
To change the order of these is trivial in the code, but I'm not sure what the desired behavior is.

**Comment by JoshRosen** [comment]
Good question. Let me cc @sungjuly, who submitted the original PR to add the `preactions` option.

At the time that the pre-actions feature was added, I believe that `spark-redshift` was automatically managing its own temporary staging table whose name was non-deterministic (it contained either a UUID or some other source of randomness) and was not performing the entire load within a transaction, so I don't think that your proposed solution would have worked with that implementation.

In terms of your actual use-case, would your needs be met by a `truncate` save-mode, as has been proposed in #213?

**Comment by sungjuly** [comment]
Originally my intention is deleting data before loading when I was submitting that feature to `spark-redshift`. That issue can be solved if those two method ordering is changed as you mentioned in this thread.

In terms of your actual use-case, I don't think your way is not enough in the real word. You will need to check data quality soon (count, threshold, etc), I would like to recommend to create your method using jdbc driver. I feel like the `spark-redshift` wants to focus on loading data to Redshift only. (not entire data flow) @JoshRosen

**Comment by JoshRosen** [comment]
I agree with @sungjuly's point that `spark-redshift`, in its current form, is insufficient for certain advanced use-cases. A similar point was raised in #197, an issue requesting the ability to execute different parts of the `COPY` path asynchronously.

It seems like both this issue and #197 might be addressed by a modularization of `spark-redshift`'s internals into a number of smaller utility components which could be composed by an end-user to implement more advanced workflows.

**Comment by bryan0** [comment]
You can do a TRUNCATE save with preactions using TRUNCATE table; ANALYZE(?); in append mode.  This doesn't really work for our use case though, because of how redshift handles truncation transactions (it autocommits), and we want the original table to still be available during the load.  But if #213 can handle this case that would be fine. 

Moving preactions before dbtable create would be much easier though and I *think* makes sense logically.

**Comment by sungjuly** [comment]
@JoshRosen I would like to say that `spark-redshift` can be handled/exposed more utility class for Redshift, that's so powerful for advanced users. I can help you if you want. Currently I'm heavy users of Redshift.

**Comment by mrakshay** [comment]
Is this issue resolved ?? Because I too facing the same issue where I want to run set of queries in "postactions" but they are not executing ??
